realtek: hpe 1920 48g PoE: mode led, mode button

evs93 opened the pull request:
JG928A has an RTL8231 on the aux mdio bus.
Add it to dts.

Add gpio fan array to dts.

Add hwmon gpio fan kernel module to DEVICE_PACKAGES.

Of note, this does not control all fans for the unit. The power supply fans are not controlled.

evs93 commented:
@svanheule here's the changes required to have fan speed on JG928A.

This has been tested by booting an initramfs on JG928A.

Thanks in advance! =)

svanheule commented:
It would be good to know how much of this caries over to the JG927A, so perhaps these changes can go into the base dtsi for JG927A/JG928A. Looking back at #15763, there were still some unsupported LEDs etc.

@janh @howels @Hurricos could perhaps also test and provide feedback.

svanheule commented:
Looking at the [HPE 1910/1920 u-boot sources](https://svanheule.net/switches/gpl_source_drops), the JG928A should have the mode LED on the RTL8231 at gpio 8.
See hpe-1910-uboot.tar.gz : v2r1_copyright/board/smb/hp1910_public/smb_public.c

evs93 commented:
> It would be good to know how much of this caries over to the JG927A, so perhaps these changes can go into the base dtsi for JG927A/JG928A. Looking back at #15763, there were still some unsupported LEDs etc.

Thanks for the feedback. I do have a JG927A to test on.

I'll have another go at checking whether JG927A has an RTL8231 on the aux mdio bus. It would be good to have another person do a sanity check there. 

At least with the old RTL8231 gpio driver I couldn't find an RTL8231 on any address on JG927A.

> Looking at the [HPE 1910/1920 u-boot sources](https://svanheule.net/switches/gpl_source_drops), the JG928A should have the mode LED on the RTL8231 at gpio 8.
> See hpe-1910-uboot.tar.gz : v2r1_copyright/board/smb/hp1910_public/smb_public.c

I'll have a go confirming that on hardware. Then I'll have a go doing LED's properly?

svanheule commented:
> At least with the old RTL8231 gpio driver I couldn't find an RTL8231 on any address on JG927A.

The bootloader sources also seems to indicate this. It will read a board ID from ID pins (internal GPIO 8:3) and only initialize an RTL8231 (for more ID pins) on the JG928A.

> I'll have a go confirming that on hardware. Then I'll have a go doing LED's properly?

Now that you're adding the RTL8231, I think it's good to get as much done as we can find out now.

evs93 commented:
Second commit has been pushed for button/LED. Going to have to have a look at JG927A later.

I also found the mode button?
I haven't figured out how the RPS led is set.

IDK whether one needs to things to device packages? kmod-button-hotplug kmod-input-core kmod-input-gpio-keys ? Similar with the LED's?

svanheule commented:
> I haven't figured out how the RPS led is set.

I realized I had a JG926A in storage, which also has an RPS LED. I probed the board a bit to figure out where it came from. It turns out the PSU has a RPS pin, which is buffered through two transistors and eventually routed to the RPS LED _and_ the PoE MCU as an active-low signal. AFAICT there is no sensing GPIO on the JG926A's SoC or RTL8231 to detect the RPS indication independently.

evs93 commented:
> I realized I had a JG926A in storage, which also has an RPS LED. I probed the board a bit to figure out where it came from. It turns out the PSU has a RPS pin, which is buffered through two transistors and eventually routed to the RPS LED _and_ the PoE MCU as an active-low signal. AFAICT there is no sensing GPIO on the JG926A's SoC or RTL8231 to detect the RPS indication independently.

Awesome. Thanks for the prompt. I guess I'm opening mine up and buzzing it out when I have a chance =)

svanheule commented:
Did some more buzzing on the JG926A's RTL8231 GPIOs:
- GPIO 7, 8, 9: Appear to be some kind of board ID. Strapped to ground on the PoE board
- GPIO 10: output to PoE board, goes to MCU
- GPIO 11: input from PoE board
- GPIO 12: output to PoE board, goes to MCU

You will probably want to verify this on your board before trying to find out their function.

svanheule commented on target/linux/realtek/dts/rtl8393_hpe_1920-48g-poe.dts:
```diff
@@ -16,6 +16,25 @@
+
+		mode_led {
+			label = "green:mode";
+			gpios = <&gpio1 8 GPIO_ACTIVE_HIGH>;
```
You should preferably use the "function" and "color" properties instead of label.
See https://mjmwired.net/kernel/Documentation/devicetree/bindings/leds/common.yaml

In absence of an appropriate LED_FUNCTION_* macro, you can just use a string (e.g `function = "mode"`)

See `default-state` and `linux,default-trigger` to set the default state/trigger instead of using the 01_leds script.

svanheule commented on target/linux/realtek/base-files/etc/board.d/01_leds:
```diff
@@ -6,6 +6,9 @@ board=$(board_name)
 
 case $board in
+	hpe,1920-48g-poe)
+		ucidef_set_led_default "mode" "MODE" "green:mode" "1"
```
This is deprecated in favour of DT properties.

svanheule reviewed (changes requested):
All changes are adding additional GPIO descriptions, so I don't mind if they are folded into one commit.

To enable use of the button, you normally don't need any additional kmod packages. By default the required kernel module is already included.

Neustradamus commented:
Nice PR :)

svanheule commented:
I reworded the fan commit a bit and merged it to main. Just the button and LED commit remaining then.

Neustradamus commented:
@svanheule: What is the status of this ticket?

svanheule commented:
The last patch is still open, which has some unresolved comments on the LED config.

evs93 commented:
> @evs93 would you mind updating the patch and addressing the LED review comments? I have a JG928A and would love to test this. Thank you.

Thanks for the prompt haha.  Knowing there's someone who wants to use it is a motivating factor. Are you just after Mode button and LED? Out of interest, what's the application?

 All the other reverse engineering work mentioned in the thread is a job where I need to reserve a lot of desk space so probably not happening until way later in the year.

evs93 committed:
realtek: hpe 1920 48g poe: add mode button, led

The JG928A has a mode button and mode LED.
Add to dts.

This is used by factory to indicate copper LED
link mode.

Factory default LED behaviour:
->On for indicating link activity / speed.
->Flashing is PoE status. (Unimplemented)

Set default timer led pattern behaviour
to same as factory.

Signed-off-by: Evan Jobling <evan@jobling.au>

evs93 commented:
@hailfinger , @svanheule when you have a moment your feedback and re-review would be appreciated.
